keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -257,27 +255,44 @@ Metadata:
   finalizers:
   - cosi.io/finalizer [3]
 spec:
-  bucketClassName: [4]
-  bucketRef: [5]
+  provisioner: [4]
+  releasePolicy: [5]
+  accessModes: [6]
+  supportedProtocols: [7]
+  bucketClassName: [8]
+  bucketRef: [9]
     name:
     namespace:
-  secretRef: [6]
+    uuid:
+    resourceVersion:
+  secretRef: [10]
     name:
     namespace:
-  objectBucketName: [7]
+  bucketIdentifier: [11]
+  parameters: 
 status:
-  bucketAttributes: <map[string]string> [8]
-  phase: [9]
+  message: [12]
+  phase: [13]
   conditions:
 ```
 1. `name`: Generated in the pattern of `<BUCKET-CLASS-NAME>'-'<RANDOM-SUFFIX>`. 
 1. `labels`: COSI controller adds the label to its managed resources for easy CLI GET ops.  Value is the driver name returned by GetDriverInfo() rpc. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
 1. `finalizers`: COSI controller adds the finalizer to defer Bucket deletion until backend deletion ops succeed.
+1. `provisioner`: The name of the drive as defined in the BucketClass.  Used by sidecars to filter BucketContents.
+1. `releasePolicy`: Prescribes outcome of a Delete events. **Note:** In Brownfield and Static cases, *Retain* is mandated.
+    - _Delete_:  the bucket and its contents are destroyed
+    - _Retain_:  the bucket and its data are preserved with only abstracting Kubernetes being destroyed
+1. `accessModes`: (Optional) Declares the level of access given to credentials provisioned through this class.     If empty, drivers may set defaults.
+1. `supportedProtocols`: (Optional) An array of protocols the associated object store supports (e.g. swift, s3, gcs, etc.). *Only* serves a descriptive purpose and is not verified.
 1. `bucketClassName`: Name of the associated `BucketClass`.
 1. `bucketRef`: the name & namespace of the associated `Bucket`.
+    - `name` : the Bucket’s name
+    - `namespace`: the Bucket’s namespace
+    - `uuid`: the Bucket’s API server generated UUID
+    - `resourceVersion`: the Bucket’s 
 1. `secretRef`: the name and namespace of the source secret in the `Bucket`'s namespace.  This `secret` may be generated by the driver or created by an admin.
-1. `objectBucketName`: the name of the actual bucket in the object store.
-1. `bucketAttributes`: stateful data relevant to the managing of the bucket but potentially inappropriate user knowledge (e.g. user's IAM role name).
+1. `bucketIdentifier`: a unique set of identifying information 
```

Comment:
Design considerations:
The id of a bucket may be more complex than a simple string name or url endpoint. SDKs often expect separate args for the name, port, region and whatever else they embed into a bucket url. An app should not have to decompose a full bucket endpoint (url) so that it can call SDK access functions. A full endpoint seems to be only useful for debugging / curl calls, outside of the app.

COSI also should not have to dig into an opaque type and figure out where the bucket name is, etc. in order to compose a secret or file content to be ingested by the app pod. Thus, in order to generate endpoint related fields and a secret, a map of these values makes sense.

Drivers also need to save some state info (eg. IAM user) so that they can de-provision a bucket that they created, and the BucketContent is a reasonable resource for that. All "state" data is ignored by COSI other than storing it in the BucketContent and passing it to the driver. The contents of the saved state is not accessed by COSI.

`objectBucketName: <string>` should be its own field in BucketContent and simply names the bucket and nothing else. It is likely redundant with one of the key-values provided in `bucketEndpointAttributes` below.
`bucketEndpointAttributes: <map[string]string>` contains all of the values needed to fully define the bucket's endpoint, eg BUCKET=, REGION=, PORT=, etc. These keys and values are provisioner-specific and are added to the secret.
`provisionerState: <base64 string blob>` contains provisioner-specific state values. Eg the IAM user may be needed to de-provision a bucket and thus would be contained in this field. COSI never looks at this value.

Reply:
The bucketEndpointAttributes would also have to be returned to the driver.  That's analogous to the `bucketIdentifier` Saad had mentioned.